walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,186 @@
 
 package com.google.sps;
 
+import com.google.sps.Event;
+import com.google.sps.Events;
+import com.google.sps.TimeRange;
+import com.google.sps.MeetingRequest;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+    Collection<TimeRange> result = new ArrayList<>(); // the return value: just a list of time ranges.
+    Collection<String> attendees = request.getAttendees();
+    Collection<String> optAttendees = request.getOptionalAttendees();
+    Iterator<Event> itr = events.iterator();
+
+    //noOptionsForTooLongOfARequest()
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+        return result;
+    }
+    // noConflicts(), optionsForNoAttendees()
+    if(events.isEmpty()){
+        TimeRange range = TimeRange.WHOLE_DAY;
+        result.add(range);
+    }
+    //eventSplitsRestriction()
+    else if(events.size() == 1){
+        Event event = itr.next();
+        Set<String> eventAttendees = new HashSet<>(event.getAttendees());
+
+        if(attendees.containsAll(eventAttendees) == false){
+            result.add(TimeRange.WHOLE_DAY);
+        }
+        else{
+            result.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, event.getWhen().start(),false));
+            result.add(TimeRange.fromStartEnd(event.getWhen().end(), TimeRange.END_OF_DAY,true));
+        }
+    }
+    else if(events.size() > 1){
+        if(optAttendees.isEmpty()){
```

Comment:
Try moving some of this logic into helper functions.  It would make it much more readable and reduce the complexity if you split it up appropriately